src/uu/tail/src/tail.rs

```diff
@@ -356,49 +653,532 @@ pub fn uu_app<'a>() -> Command<'a> {
+        self.map.get_mut(path).unwrap().reader = None;
+    }
+
+    /// Reopen the file at the monitored `path`
+    fn update_reader(&mut self, path: &Path) -> UResult<()> {
+        assert!(self.map.contains_key(path));
+        // BUG:
+        // If it's not necessary to reopen a file, GNU's tail calls seek to offset 0.
+        // However we can't call seek here because `BufRead` does not implement `Seek`.
+        // As a workaround we always reopen the file even though this might not always
+        // be necessary.
+        self.map.get_mut(path).unwrap().reader = Some(Box::new(BufReader::new(File::open(&path)?)));
+        Ok(())
+    }
+
+    /// Reload metadata from `path`, or `metadata`
+    fn update_metadata(&mut self, path: &Path, metadata: Option<Metadata>) {
+        assert!(self.map.contains_key(path));
+        self.map.get_mut(path).unwrap().metadata = if metadata.is_some() {
+            metadata
+        } else {
+            path.metadata().ok()
+        };
+    }
+
+    /// Read `path` from the current seek position forward
+    fn read_file(&mut self, path: &Path, buffer: &mut Vec<u8>) -> UResult<bool> {
+        assert!(self.map.contains_key(path));
+        let mut read_some = false;
+        let pd = self.map.get_mut(path).unwrap().reader.as_mut();
+        if let Some(reader) = pd {
             loop {
-                let mut datum = vec![];
-                match reader.read_until(b'\n', &mut datum) {
+                match reader.read_until(b'\n', buffer) {
                     Ok(0) => break,
                     Ok(_) => {
                         read_some = true;
-                        if i != last {
-                            println!("\n==> {} <==", filename);
-                            last = i;
-                        }
-                        stdout
-                            .write_all(&datum)
-                            .map_err_context(|| String::from("write error"))?;
                     }
                     Err(err) => return Err(USimpleError::new(1, err.to_string())),
                 }
             }
         }
+        Ok(read_some)
+    }
 
-        if pid_is_dead {
-            break;
+    /// Print `buffer` to stdout
+    fn print_file(&self, buffer: &[u8]) -> UResult<()> {
+        let mut stdout = stdout();
+        stdout
+            .write_all(buffer)
+            .map_err_context(|| String::from("write error"))?;
+        Ok(())
+    }
+
+    /// Read new data from `path` and print it to stdout
+    fn tail_file(&mut self, path: &Path, verbose: bool) -> UResult<bool> {
+        let mut buffer = vec![];
+        let read_some = self.read_file(path, &mut buffer)?;
+        if read_some {
+            if self.needs_header(path, verbose) {
+                self.print_header(path, true);
+            }
+            self.print_file(&buffer)?;
+
+            self.last = Some(path.to_path_buf()); // TODO: [2022-05; jhscheer] add test for this
```

Comment:
What do you still want to add a test for? There's no warning from the CI that this line is not covered by the test, so we should be good right?